core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+#ifndef TPOOL_H_
+#define TPOOL_H_
+
+#include "TCollection.h"
+#include "TObjArray.h"
+#include "ROOT/TSeq.h"
+
+template<class subc>
+class TPool {
+public:
+   explicit TPool(){};
+   explicit TPool(size_t nThreads){};
+
+   // // Map
+   // //these late return types allow for a compile-time check of compatibility between function signatures and args,
+   // //and a compile-time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   template<class F, class T> auto Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >;
```

Comment:
The return type is already calculated. I find it a lot simpler to use a new template parameter with default argument to SFINAE this away, instead of adding complexity to the return type specification.